src/cc/api/BPF.cc

```diff
@@ -619,24 +638,55 @@ StatusTuple BPF::detach_perf_event_all_cpu(open_probe_t& attr) {
 }
 
 StatusTuple USDT::init() {
-  ::USDT::Context ctx(binary_path_);
-  if (!ctx.loaded())
+  std::unique_ptr<::USDT::Context> ctx;
+  if (pid_) {
+    ctx.reset(new ::USDT::Context(pid_));
+  } else {
+    ctx.reset(new ::USDT::Context(binary_path_));
+  }
+  if (!ctx->loaded())
     return StatusTuple(-1, "Unable to load USDT " + print_name());
-  auto probe = ctx.get(name_);
-  if (probe == nullptr)
+
+  auto&& probes = ctx->get_all(binary_path_, provider_, name_);
+  if (probes.empty())
     return StatusTuple(-1, "Unable to find USDT " + print_name());
 
-  if (!probe->enable(probe_func_))
-    return StatusTuple(-1, "Failed to enable USDT " + print_name());
+  uint64_t argfmt = probes[0]->usdt_getargfmt();
+  int maxfmtidx = 0;
+  int count = 0;
+  for (auto probe : probes) {
+    if (!probe->enable(probe_func_))
+      return StatusTuple(-1, "Failed to enable USDT " + print_name());
+    // We check if the format is compatible between them
+    uint64_t tmpfmt = probe->usdt_getargfmt();
+    if (!::USDT::Probe::encloses_argfmt(argfmt, tmpfmt)) {
+      // Check if tmpfmt encloses argfmt
+      if (::USDT::Probe::encloses_argfmt(tmpfmt, argfmt)) {
+        maxfmtidx = count;
```

Comment:
Could you add some comments here? It is hard to understand what is the intention here.

Reply:
Will do. The intention, and I am not sure if this is 100% correct, is to pick the representation of the generated functions that is the most generic. Basically, the one that matches in signedness and has the largest size argument (prefer u64 over u32 for example).